engine/consensus/approvals/approvals_lru_cache.go

```diff
@@ -0,0 +1,98 @@
+package approvals
+
+import (
+	"sync"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// LruCache is a wrapper over `simplelru.LRUCache` that provides needed api for processing result approvals
+// Extends functionality of `simplelru.LRUCache` by introducing additional index for quicker access.
+type LruCache struct {
+	lru  simplelru.LRUCache
+	lock sync.RWMutex
+	// secondary index by result id, since multiple approvals could
+	// reference same result
+	byResultID map[flow.Identifier]flow.IdentifierList
+}
+
+func NewApprovalsLRUCache(limit uint) *LruCache {
+	byResultID := make(map[flow.Identifier]flow.IdentifierList)
+	// callback has to be called while we are holding lock
+	lru, _ := simplelru.NewLRU(int(limit), func(key interface{}, value interface{}) {
+		approval := value.(*flow.ResultApproval)
+		delete(byResultID, approval.Body.ExecutionResultID)
```

Comment:
⚠️ I don't understand this. It seems to me that if an approval is evicted, we will remove the index of all approvals of that result. I'm concerned this might cause inconsistent index.

Instead, I think we should just remove a single index of that approval. 

We could verify the behavior with such tests:
if the limit is 10, and if we add 11 approvals for the same result, then `TakeByResultID` should return the last 10 approvals

Reply:
You are absolutely correct, made a mistake here.